Run on cgroup v2 host

markusboehme opened the pull request:
**Issue number:** #75 


**Description of changes:** This PR aims to make the admin container work on Bottlerocket hosts with cgroup v2/the unified cgroup hierarchy enabled (rather than the current cgroup v1 default). The current version of the admin container is hindered by systemd v219 inside its image not supporting cgroup v2, and no more recent version of systemd being available for the Amazon Linux 2 base image.

The PR introduces a dedicated but unused cgroup v1 hierarchy just for the admin container. It instructs the user systemd process inside the container to use this separate hierarchy via downstream patches. In order to apply those systemd needs to be recompiled during the build process.

The core of this goes back to @bcressey noting the cgroup v1 filesystem can be mounted without any controllers. I take responsibility for actually running with this and deciding to patch systemd inside the container. The benefit of this compared to doing it on the host is that all code required for the admin container actually stays inside the admin container, and no changes at all can be observed on hosts that are not running the admin container.

**Testing done:** Locally with cgroup v1 and cgroup v2 VMs.

To test with cgroup v2 enabled, add the following user data snippet on boot

```
[settings.boot]
reboot-to-reconcile = true

[settings.boot.init]
"systemd.unified_cgroup_hierarchy" = ["1"]
```

or apply the corresponding settings at runtime using `apiclient set -j '{"settings": {"boot": {"init": {"systemd.unified_cgroup_hierarchy": ["1"]}}}}'` and reboot.

The container image with the changes from this PR can be used both on a cgroup v1 and a cgroup v2 host. Host access using `sudo sheltie` works just the same. The admin container can be stopped/disabled and restarted/reenabled.

I launched a few pods into a Kubernetes cluster with a node that used the cgroup v2 hierarchy. I then verified on the host that the containers were running in the cgroup v2 hierarchy.

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

markusboehme committed:
Dockerfile: create separate build stage for AL2 with development tools

Create a separate build stage for the Amazon Linux 2 image with
development tools already installed. This is preparation for a following
step which is also going to need the development tools installed for a
separate package build that can happen in parallel.

Signed-off-by: Markus Boehme <markubo@amazon.com>

markusboehme committed:
Dockerfile: rebuild systemd of the Amazon Linux 2 base image

Replace the systemd from the Amazon Linux 2 base image with recompiled
packages of the original SRPM. The resulting packages do not differ yet,
but this paves the way to add downstream patches to systemd in an
upcoming change.

Signed-off-by: Markus Boehme <markubo@amazon.com>

markusboehme committed:
systemd: accept non-standard base of cgroup hierarchy

Add two downstream patches to make systemd accept a non-standard
location as the base of its cgroup hierarchy. The `--cgroup-base` option
may now be used to tell systemd where to look instead of the canonical
`/sys/fs/cgroup` path. The latter will still be used by default.

Signed-off-by: Markus Boehme <markubo@amazon.com>

markusboehme committed:
use dedicated cgroup v1 hierarchy for systemd on cgroup v2 hosts

The admin container image is currently based on Amazon Linux 2 which
ships with systemd v219. This version of systemd only supports cgroup
v1. Since a user instance of systemd is run in the admin container, this
makes the admin container incompatible with a host that chose to use
cgroup v2 and will show in systemd refusing to start.

The problem can be worked around by making a cgroup v1 hierarchy
available to systemd in the admin container. This hierarchy will be
separate from the cgroup v2 hierarchy of the host, and only be mounted
inside the admin container. The user systemd process inside the
container is instructed to use that hierarchy via the downstream
`--cgroup-base` option. The start script takes care to not actually
mount any cgroup controllers, so the host-side systemd can manage
processes without interference.

An admin container running on a cgroup v1 host will not experience any
changes. An admin container running on a cgroup v2 host will have a
couple extra mounts under `/.bottlerocket/admin-container-cgroup` in its
mount namespace.

Signed-off-by: Markus Boehme <markubo@amazon.com>

markusboehme commented:
Elaborated on how I tested the change.

markusboehme commented:
In my built on aarch64, the new container image weighs in at 417 MB vs 324 MB as reported by `docker image ls`. A quick tally over the respective images' root directories comes out roughly equal, so I suspect there might be some whiteouts from upper image layers. Potentially the old files of the original systemd RPMs?

markusboehme commented:
There is an (experimental) `--squash` option to `docker build` to merge image layers and remove redundant files. This does have the desired effect of squeezing the overall image size to 316 MB (8 MB less than the image before this PR). Naturally this will also prevent sharing of image layers, which at least would be helpful for the Amazon Linux 2 base image being shared between the standard admin and control container images.

Another quick win can be had by bind-mounting the newly built RPMs from a previous build stage rather than temporarily copying them in. This gets the image down to 372 MB (45 MB saved, but still an unnecessary 48 MB heavier than before this PR) without the drawbacks of squashing the layers.

I'll update the PR with the bind-mount approach. Perhaps somebody has another idea to squeeze the image some more.

bcressey commented on Dockerfile:
```diff
@@ -1,5 +1,14 @@
-FROM public.ecr.aws/amazonlinux/amazonlinux:2 as builder
+################################################################################
+# Base image for all builds
+
+FROM public.ecr.aws/amazonlinux/amazonlinux:2 as builder-base
```
nit: typo in commit message, "sparate build stage"

bcressey commented on Dockerfile:
```diff
@@ -48,6 +48,35 @@ RUN mkdir -p /usr/share/licenses/bash && \
+WORKDIR /root/rpmbuild/SOURCES
+COPY systemd-patches/*.patch ./
+
+WORKDIR /root/rpmbuild/SPECS
```
nit: I have a bit of a reflexive dislike of building RPMs as root - can we use a throwaway unprivileged user account for this instead?

markusboehme replied:
Good point. Added a separate unprivileged `builder` user for those steps that don't need to run as root.

bcressey commented on Dockerfile:
```diff
@@ -48,6 +48,35 @@ RUN mkdir -p /usr/share/licenses/bash && \
+WORKDIR /root/rpmbuild/SPECS
+# Recreate the spec file from three parts: everything up until the last upstream
+# patch, downstream patches, everything else.
+RUN last_patch=$(grep -En '^Patch[0-9]+' systemd.spec | tail -1 | cut -f1 -d:); \
```
maybe this instead of the `grep` / `tail` / `cut` pipeline?
```
 awk '/^Patch[0-9]+/ { line=NR } END { print line }' systemd.spec
```

markusboehme replied:
I love a good awk one-liner, and this one makes it much clearer what's happening. Thanks!

bcressey commented on start_admin.sh:
```diff
@@ -212,7 +212,34 @@ install_proxy_profile
+
+# cgroup v2 compatibility crimes: systemd 219 in the admin container only
+# supports cgroup v1.
+if grep ' /sys/fs/cgroup cgroup2 ' /proc/self/mounts; then
```
If we have `findmnt`, this would be more precise:
```
findmnt /sys/fs/cgroup -o FSTYPE -n
```

bcressey commented on start_admin.sh:
```diff
@@ -212,7 +212,34 @@ install_proxy_profile
+    #
+    # In either case there is no need to populate the file systems with any
+    # actual cgroup controllers, since the host is expected to manage processes.
+    readonly cgroup_base=/bottlerocket-admin/cgroup
```
nit: perhaps `/.bottlerocket/cgroup` to "hide" the mount with other host container specific files

markusboehme replied:
Good call-out, I had forgotten about `/.bottlerocket`. I chose the more verbose `/.bottlerocket/admin-container-cgroup` as the root.

bcressey commented on start_admin.sh:
```diff
@@ -212,7 +212,34 @@ install_proxy_profile
+    mount -t tmpfs -o nosuid,nodev,noexec tmpfs "${cgroup_base}"
+    mkdir -p "${cgroup_base}"/systemd
+    mount -o remount,ro "${cgroup_base}"
+    mount -t cgroup -o none,name=systemd cgroup "${cgroup_base}"/systemd
```
When this mount is added, all processes on the system will have it appear in `/proc/<PID>/cgroup`. Does the entry go away if the admin container is subsequently disabled and this is unmounted?

markusboehme replied:
Now this is a bit peculiar. It actually does not, and does not match my expectations (e.g. compare with performing the equivalent steps via `sudo unshare -m`). I still need to dig into this some more. It seems to relate to how the admin container is disabled (e.g. if systemd exits voluntarily before entering its main loop, the mount is cleaned up). So far I could not find any process remaining in the admin container mount namespace and don't understand why the cgroup root seems to remain.

markusboehme replied:
I understand this better now. When systemd goes to remove the cgroup of the `sshd.service`, its call to `unlinkat` returns `EBUSY`. The cgroup is never removed. The cgroup root lingers on as long as there's still a cgroup in it, even if it's empty. Even if there doesn't seem to be a way to get back into the mount namespace. Aside from figuring out why removing the cgroup fails, this could also be cleaned up later by remounting the cgroup root with the same name--it'll get reused--and reaping empty cgroups.

While looking at this I found something I had completely overlooked so far, however, and I'll likely have to scrap the entire implementation. :disappointed: See my comment in the main discussion below.

markusboehme replied:
An update on the lingering cgroup root. There are two issues at play here:

1. The empty cgroup for `sshd.service` not being removed on exit. The reason for this turns out to be #62. When systemd fails to start `exit.target`, it makes a dash for the exit. Fixing this issue leads to a proper exit and all cgroups being removed.
1. A much more interesting timing-dependent kernel behavior. I'm dedicating the rest of the comment to this one.

Unmounting a cgroup file system can trigger removal of a cgroup root if it was the last mount of that root and the root is otherwise unused. Since with 1. out of the way all cgroups are removed before the container exits, leaving its mount namespace should trigger removal of the cgroup root. However, since the removal of an individual cgroup is an asynchronous multi-step endeavor, unmount and cgroup removal can actually race, with the unmount happening before a cgroup has been fully removed. Once the `rmdir` syscall returns, user-visible effects of cgroup removal have occured, but internal bookkeeping tasks are not yet complete and happen asynchronously. In this case, the cgroup root lingers on and can be observed in e.g. `/proc/1/cgroup` even after the admin container exited.

The issue can be trivially reproduced with e.g. the following sequence of commands (note that `rmdir` and `umount` will need to happen in short succession, so it's best to run this as a script):

```
mkdir -p /mnt/cgroot
mount -t cgroup -o none,name=cgroot cgroup /mnt/cgroot
mkdir /mnt/cgroot/cg
rmdir /mnt/cgroot/cg
umount /mnt/cgroot
```

I spent some time looking at fixing this from the kernel side, but the highly asynchronous nature of the cgroup release code makes this a delicate matter with plenty opportunity to get things wrong. Knowing what to look for now, I noticed I haven't been the first one to observe or try to fix this one either. [The discussion on the cgroups mailing list](https://lore.kernel.org/lkml/20220404142535.145975-1-minhquangbui99@gmail.com/) concluded not to fix this race.

For Bottlerocket's purposes, this is a cosmetic issue. Nothing is really being leaked. The same cgroup root is available and gets reused when the admin container is started again. I'd therefore propose to just live with that for the time being (a workaround would be to re-mount/unmount the hierarchy after some time, e.g. in a `host-ctr` clean-up hook).

bcressey reviewed (approved):
Nice work. It's an ugly thing to have to do, but you found a clean way to do it.

bcressey commented:
> I'll update the PR with the bind-mount approach. Perhaps somebody has another idea to squeeze the image some more.

systemd isn't in the base container image; it comes in from when the "Development Tools" group is installed. If the layers are structured so that we don't install that group until the end - or potentially not at all, if we replace it with a more targeted package install command - then we could avoid the whiteout of the previous systemd install.

bcressey commented:
Oh, there's already a more targeted `yum install` - I think we just need to install the RPMs from the bind mount before rather than after that point.

markusboehme committed:
Dockerfile: split packages to install onto individual lines

Splitting packages to install using yum onto individual, sorted lines to
facilitate further additions. No semantic changes.

Signed-off-by: Markus Boehme <markubo@amazon.com>

markusboehme commented:
>> I'll update the PR with the bind-mount approach. Perhaps somebody has another idea to squeeze the image some more.
>
> systemd isn't in the base container image; it comes in from when the "Development Tools" group is installed. If the layers are structured so that we don't install that group until the end - or potentially not at all, if we replace it with a more targeted package install command - then we could avoid the whiteout of the previous systemd install.

True, I jumped to conclusions here, thanks for noticing! systemd is not in the base image, but actually pulled in as a dependency of the `openssh-server` package. This means we can cut the dead weight by making sure the downstream systemd build is installed alongside the other packages baked into the admin container image.

markusboehme commented:
The earlier force push addresses feedback by @bcressey. Leaving the PR in draft mode until I figure out why the cgroup root lingers on when the admin container is disabled.

markusboehme commented:
What I had completely overlooked so far is that the admin container may be run without superpowers (`superpowered = false` in the user data). It then will not have the necessary permissions to mount a cgroup v1 root by itself in `start_admin.sh`. While there's other problems in running without superpowers at the moment (#78), this probably tanks the approach of containing all the compatibility work in the admin container.

markusboehme commented:
I will merge this tomorrow unless someone speaks up against it. Consensus on #78 is to not worry about the "admin container without superpowers" use case, since it's been broken for a while already without anyone noticing and it not being super useful if it weren't.

bcressey reviewed (approved):
Nicely done!
